// react-native-render-html/lib/commonjs/elements/useIMGElementState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useIMGElementState;

var _react = require("react");

var _reactNative = require("react-native");

var _useImageNaturalDimensions = _interopRequireDefault(require("./useImageNaturalDimensions"));

var _useImageConcreteDimensions = _interopRequireDefault(require("./useImageConcreteDimensions"));

var _defaultInitialImageDimensions = _interopRequireDefault(require("./defaultInitialImageDimensions"));

var _getIMGState = require("./getIMGState");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getImageSizeAsync({
  uri,
  headers
}) {
  return new Promise((onsuccess, onerror) => {
    const onImageDimensionsSuccess = (width, height) => onsuccess({
      width,
      height
    });

    if (headers) {
      _reactNative.Image.getSizeWithHeaders(uri, headers, onImageDimensionsSuccess, onerror);
    } else {
      _reactNative.Image.getSize(uri, onImageDimensionsSuccess, onerror);
    }
  });
}

function useFetchedNaturalDimensions(props) {
  const {
    source,
    cachedNaturalDimensions
  } = props;
  const {
    error,
    flatStyle,
    naturalDimensions,
    specifiedDimensions,
    onError,
    onNaturalDimensions
  } = (0, _useImageNaturalDimensions.default)(props);
  const hasCachedDimensions = !!cachedNaturalDimensions;
  (0, _react.useEffect)(function fetchPhysicalDimensions() {
    let cancelled = false;

    if (source.uri && !hasCachedDimensions) {
      getImageSizeAsync({
        uri: source.uri,
        headers: source.headers
      }).then(dimensions => !cancelled && onNaturalDimensions(dimensions)).catch(e => !cancelled && onError(e || {}));
      return () => {
        cancelled = true;
      };
    }
  }, [source.uri, source.headers, onNaturalDimensions, onError, hasCachedDimensions]);
  return {
    specifiedDimensions,
    flatStyle,
    naturalDimensions,
    error,
    onError,
    onNaturalDimensions
  };
}
/**
 * This hook will compute concrete dimensions from image natural dimensions and
 * constraints. It will fetch the image and get its dimensions.
 *
 * @remarks If you know the dimensions beforehand, use
 * {@link useIMGElementStateWithCache} instead to save a network request and
 * prevent a layout shift.
 */


function useIMGElementState(props) {
  const {
    alt,
    altColor,
    source,
    contentWidth,
    computeMaxWidth,
    objectFit,
    initialDimensions = _defaultInitialImageDimensions.default
  } = props;
  const {
    naturalDimensions,
    specifiedDimensions,
    flatStyle,
    onError,
    error
  } = useFetchedNaturalDimensions(props);
  const concreteDimensions = (0, _useImageConcreteDimensions.default)({
    flatStyle,
    naturalDimensions,
    specifiedDimensions,
    computeMaxWidth,
    contentWidth
  });
  return (0, _getIMGState.getIMGState)({
    error,
    alt,
    altColor,
    concreteDimensions,
    containerStyle: flatStyle,
    initialDimensions,
    objectFit,
    onError,
    source
  });
}
