// react-native-render-html/lib/commonjs/hooks/useAssembledCommonProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useAssembledCommonProps;

var _mergeCollapsedMargins = _interopRequireDefault(require("../helpers/mergeCollapsedMargins"));

var _RenderRegistryProvider = require("../context/RenderRegistryProvider");

var _SharedPropsProvider = require("../context/SharedPropsProvider");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @internal
 */
function useAssembledCommonProps({
  tnode,
  propsFromParent,
  sharedProps
}, TDefault) {
  var _ref;

  const {
    Default,
    Custom
  } = (0, _RenderRegistryProvider.useRendererConfig)(tnode);
  const containerProps = (0, _SharedPropsProvider.useDefaultContainerProps)();
  const assembledProps = {
    tnode,
    propsFromParent,
    sharedProps,
    TDefaultRenderer: TDefault,
    style: (0, _mergeCollapsedMargins.default)(propsFromParent === null || propsFromParent === void 0 ? void 0 : propsFromParent.collapsedMarginTop, tnode.getNativeStyles()),
    type: tnode.type === 'text' || tnode.type === 'phrasing' ? 'text' : 'block',
    propsForChildren: tnode.tagName ? {} : propsFromParent,
    InternalRenderer: Default || TDefault,
    ...containerProps
  };
  return {
    assembledProps,
    Renderer: (_ref = Custom !== null && Custom !== void 0 ? Custom : Default) !== null && _ref !== void 0 ? _ref : TDefault
  };
}
